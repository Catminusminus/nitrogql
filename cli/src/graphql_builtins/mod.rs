use crate::ast::{
    base::{Ident, Keyword, Pos},
    r#type::{NamedType, NonNullType, Type},
    type_system::{
        ArgumentsDefinition, DirectiveDefinition, InputValueDefinition, ScalarTypeDefinition,
        TypeDefinition, TypeSystemDefinitionOrExtension,
    },
    value::{StringValue, Value},
};

/// Generate built-in definitions.
/// TODO: make this configurable
pub fn generate_builtins() -> Vec<TypeSystemDefinitionOrExtension<'static>> {
    let type_definitions = vec![
        scalar("Int"),
        scalar("Float"),
        scalar("String"),
        scalar("Boolean"),
        scalar("ID"),
    ];

    let directive_definitions = vec![
        directive(
            "skip",
            vec![(
                "if",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("Boolean"),
                    }),
                })),
                None,
            )],
            vec!["FIELD", "FRAGMENT_SPREAD", "INLINE_FRAGMENT"],
        ),
        directive(
            "include",
            vec![(
                "if",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("Boolean"),
                    }),
                })),
                None,
            )],
            vec!["FIELD", "FRAGMENT_SPREAD", "INLINE_FRAGMENT"],
        ),
        directive(
            "deprecated",
            vec![(
                "reason",
                Type::Named(NamedType {
                    name: ident("String"),
                }),
                Some(Value::StringValue(StringValue {
                    position: Pos::builtin(),
                    value: String::from("No longer supported"),
                })),
            )],
            vec![
                "FIELD_DEFINITION",
                "ARGUMENT_DEFINITION",
                "INPUT_FIELD_DEFINITION",
                "ENUM_VALUE",
            ],
        ),
        directive(
            "specifiedBy",
            vec![(
                "url",
                Type::NonNull(Box::new(NonNullType {
                    r#type: Type::Named(NamedType {
                        name: ident("String"),
                    }),
                })),
                None,
            )],
            vec!["SCALAR"],
        ),
    ];

    type_definitions
        .into_iter()
        .map(|def| TypeSystemDefinitionOrExtension::TypeDefinition(def))
        .chain(
            directive_definitions
                .into_iter()
                .map(|def| TypeSystemDefinitionOrExtension::DirectiveDefinition(def)),
        )
        .collect()
}

fn scalar(name: &str) -> TypeDefinition {
    TypeDefinition::Scalar(ScalarTypeDefinition {
        description: None,
        position: Pos::builtin(),
        name: ident(name),
        directives: vec![],
        scalar_keyword: keyword("scalar"),
    })
}

fn directive<'a>(
    name: &'a str,
    arguments: Vec<(&'a str, Type<'a>, Option<Value<'a>>)>,
    locations: Vec<&'a str>,
) -> DirectiveDefinition<'a> {
    DirectiveDefinition {
        description: None,
        position: Pos::builtin(),
        name: ident(name),
        arguments: if arguments.is_empty() {
            None
        } else {
            Some(ArgumentsDefinition {
                input_values: arguments
                    .into_iter()
                    .map(|(name, ty, default_value)| InputValueDefinition {
                        description: None,
                        position: Pos::builtin(),
                        name: ident(name),
                        r#type: ty,
                        default_value,
                        directives: vec![],
                    })
                    .collect(),
            })
        },
        repeatable: None,
        locations: locations.into_iter().map(ident).collect(),
        directive_keyword: keyword("directive"),
    }
}

fn ident(name: &str) -> Ident {
    Ident {
        name,
        position: Pos::builtin(),
    }
}

fn keyword(name: &str) -> Keyword {
    Keyword {
        name,
        position: Pos::builtin(),
    }
}
