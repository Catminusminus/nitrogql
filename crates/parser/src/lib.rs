mod parser;
mod tests;

pub use parser::{parse_operation_document, parse_type_system_document, ParseError};
