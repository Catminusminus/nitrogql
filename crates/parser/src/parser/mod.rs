use nitrogql_ast::{OperationDocument, TypeSystemOrExtensionDocument};
use pest::Parser;
use pest_derive::Parser;

use self::builder::{build_operation_document, build_type_system_or_extension_document};

mod builder;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RawParser;

pub fn parse_operation_document(
    document: &str,
) -> Result<OperationDocument, pest::error::Error<Rule>> {
    let res = RawParser::parse(Rule::ExecutableDocument, document)?;

    Ok(build_operation_document(res))
}

pub fn parse_type_system_document(
    document: &str,
) -> Result<TypeSystemOrExtensionDocument, pest::error::Error<Rule>> {
    let res = RawParser::parse(Rule::TypeSystemExtensionDocument, document)?;

    Ok(build_type_system_or_extension_document(res))
}
