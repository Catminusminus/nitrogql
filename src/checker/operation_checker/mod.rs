use crate::{
    graphql_parser::ast::{
        base::HasPos,
        operations::{
            ExecutableDefinition, FragmentDefinition, OperationDefinition, OperationType,
            VariablesDefinition,
        },
        selection_set::{SelectionSet, Selection},
        type_system::{FieldDefinition, TypeDefinition},
        OperationDocument, TypeSystemDocument,
    },
};

use super::{definition_map::{DefinitionMap, generate_definition_map}, error::{CheckError, CheckErrorMessage, TypeKind}, common::check_directives, types::{inout_kind_of_type, TypeInOutKind}};

pub fn check_operation_document(
    schema: &TypeSystemDocument,
    document: &OperationDocument,
) -> Vec<CheckError> {
    let mut result = vec![];
    let definitions = generate_definition_map(schema);

    let operation_num = document
        .definitions
        .iter()
        .filter(|def| matches!(def, ExecutableDefinition::OperationDefinition(_)))
        .count();

    for (idx, def) in document.definitions.iter().enumerate() {
        match def {
            ExecutableDefinition::OperationDefinition(op) => {
                match op.name {
                    None => {
                        // Unnamed operation must be the only operation in the document
                        if operation_num != 1 {
                            result.push(
                                CheckErrorMessage::UnNamedOperationMustBeSingle
                                    .with_pos(op.position),
                            );
                        }
                    }
                    Some(ref name) => {
                        // Find other one with same name
                        let dup = document
                            .definitions
                            .iter()
                            .take(idx)
                            .find(|other| match other {
                                ExecutableDefinition::OperationDefinition(def) => {
                                    def.name.map_or(false, |n| n.name == name.name)
                                }
                                ExecutableDefinition::FragmentDefinition(_) => false,
                            });
                        if let Some(other) = dup {
                            result.push(
                                CheckErrorMessage::DuplicateOperationName {
                                    operation_type: op.operation_type,
                                }
                                .with_pos(name.position).with_additional_info(vec![(
                                    *other.position(),
                                    CheckErrorMessage::AnotherDefinitionPos { name: name.name.to_owned()  }
                                )]),
                            );
                        }
                    }
                }

                check_operation(&definitions, op, &mut result);
            }
            ExecutableDefinition::FragmentDefinition(def) => {
                // Find other one with same name
                let dup = document
                    .definitions
                    .iter()
                    .take(idx)
                    .find(|other| match other {
                        ExecutableDefinition::OperationDefinition(def) => false,
                        ExecutableDefinition::FragmentDefinition(other) => {
                            other.name.name == def.name.name
                        }
                    });
                if let Some(other) = dup {
                    result.push(
                        CheckErrorMessage::DuplicateFragmentName {
                            other_position: *other.position(),
                        }
                        .with_pos(def.name.position),
                    );
                }

                check_fragment(&definitions, def, &mut result);
            }
        }
    }
    result
}

fn check_operation(
    definitions: &DefinitionMap,
    op: &OperationDefinition,
    result: &mut Vec<CheckError>,
) {
    let root_type = {
        let root_type_name = definitions
            .schema
            .map(|schema_def| {
                schema_def
                    .definitions
                    .iter()
                    .find_map(|(key, value)| (*key == op.operation_type).then_some(value.name))
                    .ok_or(schema_def)
            })
            .unwrap_or_else(|| {
                Ok(match op.operation_type {
                    OperationType::Query => "Query",
                    OperationType::Mutation => "Mutation",
                    OperationType::Subscription => "Subscription",
                })
            });

        match root_type_name {
            Ok(root_type_name) => {
                let Some(root_type) = definitions.types.get(root_type_name) else {
                    result.push(
                        CheckErrorMessage::UnknownType { name: root_type_name.to_owned() }.with_pos(op.position)
                    );
                    return;
                };
                root_type
            }
            Err(schema_def) => {
                result.push(
                    CheckErrorMessage::NoRootType {
                        operation_type: op.operation_type,
                    }
                    .with_pos(*op.position())
                    .with_additional_info(vec![
                        (schema_def.position, CheckErrorMessage::RootTypesAreDefinedHere)
                    ])
                );
                return;
            }
        }
    };
    check_directives(definitions, &op.directives, match op.operation_type {
        OperationType::Query => "QUERY",
        OperationType::Mutation => "MUTATION",
        OperationType::Subscription => "SUBSCRIPTION",
    }, result);
    if let Some(ref variables_definition) = op.variables_definition {
        check_variables_definition(definitions, variables_definition, result);
    }
    if op.operation_type == OperationType::Subscription {
        todo!("Single root field check");
    }
    check_selection_set(
        definitions,
        op.variables_definition.as_ref(),
        root_type,
        &op.selection_set,
        result,
    );
}

fn check_fragment(
    definitions: &DefinitionMap,
    op: &FragmentDefinition,
    result: &mut Vec<CheckError>,
) {
    todo!()
}

fn check_variables_definition(
    definitions: &DefinitionMap,
    variables: &VariablesDefinition,
    result: &mut Vec<CheckError>,
) {
    let mut seen_variables = vec![];
    for v in variables.definitions.iter() {
        if seen_variables.contains(&v.name.name) {
            result.push(
                CheckErrorMessage::DuplicatedVariableName { name: v.name.name.to_owned() }
                .with_pos(v.pos)
            );
        } else {
            seen_variables.push(v.name.name);
        }
        let type_kind = inout_kind_of_type(definitions, &v.r#type);
        match type_kind {
            None => {
                result.push(
                    CheckErrorMessage::UnknownType { name: v.r#type.unwrapped_type().name.name.to_owned() }
                    .with_pos(*v.r#type.position())
                );
            }
            Some(t) if t.is_input_type() => {
            }
            _ => {
                result.push(
                    CheckErrorMessage::NoOutputType { name: v.r#type.unwrapped_type().name.name.to_owned() }
                    .with_pos(*v.r#type.position())
                );
            }
        }
    }
}

fn check_selection_set(
    definitions: &DefinitionMap,
    variables: Option<&VariablesDefinition>,
    root_type: &TypeDefinition,
    selection_set: &SelectionSet,
    result: &mut Vec<CheckError>,
) {
    let root_type_name = root_type.name().expect("Type definition must have name");
    let root_fields = direct_fields_of_output_type(root_type);
    let Some(root_fields) = root_fields else {
        result.push(
            CheckErrorMessage::SelectionOnInvalidType { kind: 
                kind_of_type_definition(root_type),
                name: root_type_name.to_owned(),
            }
                .with_pos(selection_set.position)
                .with_additional_info(vec![
                    (*root_type.position(), CheckErrorMessage::DefinitionPos { name: root_type_name.to_owned()})
                ])
        );
        return;
    };

    for selection in selection_set.selections.iter() {
        match selection {
            Selection::Field(field_selection) => {
                let target_field = root_fields.iter().find(|field| {
                    field.name.name == field_selection.name.name
                });
                let Some(target_field) = target_field else {
                    result.push(
                        CheckErrorMessage::FieldNotFound { field_name: 
                            field_selection.name.name.to_owned(),
                             type_name: root_type_name.to_owned(),
                         }.with_pos(field_selection.name.position)
                         .with_additional_info(vec![
                            (*root_type.position(), CheckErrorMessage::DefinitionPos {
                                name: root_type_name.to_owned()
                             })
                         ])
                    );
                    continue;
                };
                check_directives(definitions, &field_selection.directives, "FIELD", result);
                // todo: duplicate name check, arguments check
                if let Some(ref selection_set) = field_selection.selection_set {
                    let Some(target_field_type) = definitions.types.get(
                        target_field.r#type.unwrapped_type().name.name
                    ) else {
                        result.push(CheckErrorMessage::TypeSystemError.with_pos(selection_set.position));
                        continue;
                    };

                    check_selection_set(definitions, variables, target_field_type, selection_set, result);
                }
            }
            Selection::FragmentSpread(_) => todo!(),
            Selection::InlineFragment(_) => todo!(),
        }
    }
}

fn direct_fields_of_output_type<'a, 'src>(
    ty: &'a TypeDefinition<'src>,
) -> Option<&'a [FieldDefinition<'src>]> {
    match ty {
        TypeDefinition::Object(obj) => Some(&obj.fields),
        TypeDefinition::Interface(obj) => Some(&obj.fields),
        TypeDefinition::Union(_)
        | TypeDefinition::Scalar(_)
        | TypeDefinition::Enum(_)
        | TypeDefinition::InputObject(_) => None,
    }
}

fn kind_of_type_definition(definition: &TypeDefinition) -> TypeKind {
    match definition {
        TypeDefinition::Scalar(_) => TypeKind::Scalar,
        TypeDefinition::Object(_) => TypeKind::Object,
        TypeDefinition::Interface(_) => TypeKind::Interface,
        TypeDefinition::Union(_) => TypeKind::Union,
        TypeDefinition::Enum(_) => TypeKind::Enum,
        TypeDefinition::InputObject(_) => TypeKind::InputObject,
    }
}
