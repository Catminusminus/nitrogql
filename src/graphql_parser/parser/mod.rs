use anyhow::Result;
use pest::Parser;
use pest_derive::Parser;

use self::builder::build_operation_document;

use super::ast::OperationDocument;

mod builder;

#[derive(Parser)]
#[grammar = "graphql_parser/parser/grammar.pest"]
pub struct RawParser;

pub fn parse_operation_document(document: &str) -> Result<OperationDocument> {
    let res = RawParser::parse(Rule::ExecutableDocument, document)?;

    Ok(build_operation_document(res))
}
