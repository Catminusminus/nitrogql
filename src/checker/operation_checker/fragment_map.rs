use std::collections::HashMap;

use crate::graphql_parser::ast::{
    operations::{ExecutableDefinition, FragmentDefinition},
    OperationDocument,
};

pub type FragmentMap<'a, 'src> = HashMap<&'a str, &'a FragmentDefinition<'src>>;

pub fn generate_fragment_map<'a, 'src>(
    document: &'a OperationDocument<'src>,
) -> FragmentMap<'a, 'src> {
    document
        .definitions
        .iter()
        .flat_map(|def| {
            if let ExecutableDefinition::FragmentDefinition(fr) = def {
                Some((fr.name.name, fr))
            } else {
                None
            }
        })
        .collect()
}
