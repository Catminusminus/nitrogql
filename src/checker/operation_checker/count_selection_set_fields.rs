use crate::graphql_parser::ast::selection_set::{Selection, SelectionSet};

use super::fragment_map::FragmentMap;

pub fn selection_set_has_more_than_one_fields(
    fragment_map: &FragmentMap,
    selection_set: &SelectionSet,
) -> bool {
    selection_set_has_more_than_one_fields_impl(fragment_map, selection_set, &[]) > 1
}

fn selection_set_has_more_than_one_fields_impl(
    fragment_map: &FragmentMap,
    selection_set: &SelectionSet,
    seen_fragments: &[&str],
) -> usize {
    let mut count = 0;
    for selection in selection_set.selections.iter() {
        match selection {
            Selection::Field(_) => {
                count += 1;
            }
            Selection::FragmentSpread(fragment_spread) => {
                if seen_fragments.contains(&fragment_spread.fragment_name.name) {
                    // prevent infinite recursions
                    continue;
                }
                let fragment_def = fragment_map.get(fragment_spread.fragment_name.name);
                match fragment_def {
                    None => {
                        // This should be handled elsewhere
                        continue;
                    }
                    Some(f) => {
                        let seen_fragments: Vec<_> = seen_fragments
                            .into_iter()
                            .map(|s| *s)
                            .chain(vec![fragment_spread.fragment_name.name])
                            .collect();
                        count += selection_set_has_more_than_one_fields_impl(
                            fragment_map,
                            &f.selection_set,
                            &seen_fragments,
                        );
                    }
                }
            }
            Selection::InlineFragment(inline_fragment) => {
                count += selection_set_has_more_than_one_fields_impl(
                    fragment_map,
                    &inline_fragment.selection_set,
                    seen_fragments,
                );
            }
        }
    }
    count
}
