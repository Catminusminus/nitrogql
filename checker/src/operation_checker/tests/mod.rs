use graphql_builtins::generate_builtins;
use nitrogql_ast::TypeSystemDocument;
use nitrogql_parser::parser::parse_type_system_document;
use nitrogql_semantics::resolve_extensions;

mod operations {
    use insta::assert_debug_snapshot;

    use crate::operation_checker::check_operation_document;
    use nitrogql_ast::TypeSystemDocument;
    use nitrogql_parser::parser::parse_operation_document;

    use super::parse_to_type_system_document;

    fn type_system() -> TypeSystemDocument<'static> {
        parse_to_type_system_document(
            "
            schema {
                subscription: S
            }

            type S {
                foo: [Foo!]!
                bar: [String!]!
            }
            type Foo {
                hoge: Int!
                piyo: Int!
            }
        ",
        )
    }

    #[test]
    fn subscription_root_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            subscription a { foo { hoge piyo } }
            subscription b { foo bar }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn subscription_root_field_recursing() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            subscription { ...F }
            fragment F on S {
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }
}

mod operation_directives {
    use insta::assert_debug_snapshot;

    use crate::operation_checker::check_operation_document;
    use nitrogql_ast::TypeSystemDocument;
    use nitrogql_parser::parser::parse_operation_document;

    use super::parse_to_type_system_document;

    fn type_system() -> TypeSystemDocument<'static> {
        parse_to_type_system_document(
            "
            directive @dir_bool_nonnull(bool: Boolean!) repeatable on QUERY | FIELD
            directive @dir_bool(bool: Boolean) repeatable on MUTATION | FIELD

            type Query {
                foo: Int!
            }
            type Mutation {
                bar: Int!
            }

            directive @dir_input(input: MyInput) on QUERY | FIELD
            input MyInput {
                bool: Boolean
                int: Int!
            }
        ",
        )
    }

    #[test]
    fn unknown_directive() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @unknown_dir {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }
    #[test]
    fn wrong_location() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn wrong_argument_type() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool_nonnull(bool: 3) {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn unknown_argument() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query @dir_bool_nonnull(bool: false, another: true) {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn no_argument() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q @dir_bool_nonnull {
                foo
            }
            mutation m
              @dir_bool
              @dir_bool(bool: null)
              @dir_bool(bool: true) {
                bar
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn variable_check() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q($b: Boolean, $b2: Boolean!, $i: Int!)
                @dir_bool_nonnull(bool: $b)
                @dir_bool_nonnull(bool: $b2)
                @dir_bool_nonnull(bool: $i)
                 {
                foo
                @dir_bool(bool: $b)
                @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn unknown_variable() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q1 @dir_bool_nonnull(bool: $b) {
                foo
                @dir_bool(bool: $b)
            }
            query q2($b: Boolean!) @dir_bool_nonnull(bool: $b2) {
                foo @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn input_missing_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn input_extra_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true, int: 3 str: \"foobar\" })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn input_field_type_mismatch() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: true, int: \"foobar\" })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn input_field_null_literal() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            @dir_input(input: { bool: null, int: 3 })
            {
                foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }
}

mod selection_set {
    use insta::assert_debug_snapshot;

    use crate::operation_checker::check_operation_document;
    use nitrogql_ast::TypeSystemDocument;
    use nitrogql_parser::parser::parse_operation_document;

    use super::parse_to_type_system_document;

    fn type_system() -> TypeSystemDocument<'static> {
        parse_to_type_system_document(
            "
            type Query {
                foo: Int!
                user: User
                users(name: String): [User!]!
            }
            type User {
                id: ID!
                name: String!
                age: Int
                pet: Animal
                relations: [UserOrPost!]!
            }
            interface Animal {
                id: ID!
                name: String!
            }
            type Post {
                id: ID!
                title: String!
            }
            union UserOrPost = User | Post
        ",
        )
    }

    #[test]
    fn unknown_selected_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            {
                user
                user2
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn duplicated_selected_field() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query q1 {
                user
                user { id name }
            }
            query q2 {
                user
                alias: user # ‚Üê this is ok
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn duplicated_alias() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query
            {
                user
                user: foo
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn invalid_nested_selection() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                foo { value }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn check_field_arguments() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                users(name: \"uhyo\") { id name age }
                user(arg: 123)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn supports_typename_metafield() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    __typename
                    pet {
                        __typename
                    }
                    relations {
                        __typename
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }
}

mod fragments {
    use insta::assert_debug_snapshot;

    use crate::operation_checker::check_operation_document;
    use nitrogql_ast::TypeSystemDocument;
    use nitrogql_parser::parser::parse_operation_document;

    use super::parse_to_type_system_document;

    fn type_system() -> TypeSystemDocument<'static> {
        parse_to_type_system_document(
            "
            directive @dir_bool(bool: Boolean!) on FIELD
            scalar CustomScalar
            
            type Query {
                foo: Int!
                user: User
                users(name: String): [User!]!
                hasTitle: HasTitle!
                postOrTag: PostOrTag
                userOrSchedule: UserOrSchedule
            }
            type User {
                id: ID!
                name: String!
                age: Int
                userKind: UserKind
                followers: [User!]!
            }
            enum UserKind { NormalUser PremiumUser }
            input MyInput {
                arg: String! = \"\"
            }

            type Post implements HasTitle {
                id: ID!
                title: String!
                body: String!
            }
            interface HasTitle {
                title: String!
            }
            interface HasLabel {
                label: String!
            }

            type Tag implements HasLabel {
                id: ID!
                label: String!
            }
            type Schedule {
                id: ID!
            }
            union UserOrTag = User | Tag
            union PostOrTag = Post | Tag
            union UserOrSchedule = User | Schedule
        ",
        )
    }

    #[test]
    fn unknown_fragment_target() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user }
            fragment A on Nothing {
                id
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn invalid_fragment_target() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user }
            fragment OnScalar on CustomScalar {
                id
            }
            fragment OnEnum on UserKind {
                id
            }
            fragment OnInput on MyInput {
                arg
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_obj() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on Post {
                id
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on HasTitle {
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ...F
            }}
            fragment F on PostOrTag {
                ... on Post {
                    id
                }
                ... on Tag {
                    id
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_intf_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { hasTitle {
                ...F
                ...G
                ...H
            }}
            fragment F on HasLabel {
                label
            }
            fragment G on HasTitle {
                title
            }
            fragment H on Post {
                title
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_intf_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { hasTitle {
                ...F
                ...G
            }}
            fragment F on PostOrTag {
                ... on Post {
                    title
                }
                ... on Tag {
                    label
                }
            }
            fragment G on UserOrPost {
                ... on Post {
                    title
                }
                ... on User {
                    name
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_fragment_target_union_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { userOrSchedule {
                ...F
                ...G
            }}
            fragment F on PostOrTag {
                ... on Post {
                    title
                }
                ... on Tag {
                    label
                }
            }
            fragment G on UserOrTag {
                ... on Tag {
                    label
                }
                ... on User {
                    name
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn fragment_variables() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query($b1: Boolean!) { user {
                ...F
            }}
            fragment F on User {
                id @dir_bool(bool: $b1)
                name @dir_bool(bool: $b2)
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn inline_fragment_without_condition() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query($b1: Boolean!) { user {
                id
                ... {
                    name
                    age
                }
                ... {
                    something
                }
            }}
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc));
    }

    #[test]
    fn wrong_inline_fragment_target_obj() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query { user {
                ... on Post {
                    id
                    title
                }
            }}
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_inline_fragment_target_intf() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ... on HasTitle {
                        title
                    }
                }
                hasTitle {
                    ... on HasTitle {
                        title
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn wrong_inline_fragment_target_union() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ... on PostOrTag {
                        ... on Post {
                            id
                        }
                        ... on Tag {
                            id
                        }
                    }
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn fragment_spread_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                id
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn fragment_spread_indirect_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                id
                ...G
            }
            fragment G on User {
                age
                ...F
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }

    #[test]
    fn fragment_spread_nested_recursion() {
        let schema = type_system();
        let doc = parse_operation_document(
            "
            query {
                user {
                    ...F
                }
            }
            fragment F on User {
                followers {
                    id
                    ...F
                }
            }
        ",
        )
        .unwrap();

        assert_debug_snapshot!(check_operation_document(&schema, &doc))
    }
}

fn parse_to_type_system_document(source: &str) -> TypeSystemDocument {
    let mut doc = parse_type_system_document(source).unwrap();
    doc.extend(generate_builtins());
    let doc = resolve_extensions(doc).unwrap();
    doc
}
