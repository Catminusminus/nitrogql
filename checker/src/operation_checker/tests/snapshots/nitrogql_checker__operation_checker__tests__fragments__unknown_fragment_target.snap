---
source: checker/src/operation_checker/tests/mod.rs
expression: "check_operation_document(&schema, &doc)"
---
[
    CheckError {
        position: Pos {
            line: 2,
            column: 26,
            file: 0,
            builtin: false,
        },
        message: UnknownType {
            name: "Nothing",
        },
        additional_info: [],
    },
]
